modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/rel/CollectNode.java

```diff
@@ -0,0 +1,239 @@
+                return new ArrayCollector<>(ctx.rowHandler(), rowFactory, IN_BUFFER_SIZE);
+            case MAP:
+                return new MapCollector<>(ctx.rowHandler(), rowFactory, IN_BUFFER_SIZE);
+            default:
+                throw new RuntimeException("Unsupported collectionType: " + collectionType.getSqlTypeName());
+        }
+    }
+
+    /** */
+    private abstract static class Collector<Row> implements Supplier<Row> {
+        /** */
+        protected final RowHandler<Row> rowHandler;
+
+        /** */
+        protected final RowHandler.RowFactory<Row> rowFactory;
+
+        /** */
+        protected final int cap;
+
+        /** */
+        Collector(
+            RowHandler<Row> handler,
+            RowHandler.RowFactory<Row> factory,
+            int cap
+        ) {
+            rowHandler = handler;
+            rowFactory = factory;
+            this.cap = cap;
+        }
+
+        /** */
+        public abstract void push(Row row);
+
+        /** */
+        public abstract void clear();
+
+        /** */
+        protected abstract Object outData();
+
+        /** {@inheritDoc} */
+        @Override public Row get() {
+            Row out = rowFactory.create();
+
+            rowHandler.set(0, out, outData());
+            return out;
+        }
+    }
+
+    /** */
+    private static class MapCollector<Row> extends Collector<Row> {
+        /** */
+        private final Map<Object, Object> outBuf;
+
+        /** */
+        private MapCollector(
+            RowHandler<Row> handler,
+            RowHandler.RowFactory<Row> rowFactory,
+            int cap
+        ) {
+            super(handler, rowFactory, cap);
+            outBuf = new LinkedHashMap<>(cap);
+        }
+
+        /** {@inheritDoc} */
+        @Override protected Object outData() {
+            return Collections.unmodifiableMap(outBuf);
+        }
+
+        /** {@inheritDoc} */
+        @Override public void push(Row row) {
+            outBuf.put(rowHandler.get(0, row), rowHandler.get(1, row));
+        }
+
+        /** {@inheritDoc} */
+        @Override public void clear() {
+            outBuf.clear();
+        }
+    }
+
+    /** */
+    private static class ArrayCollector<Row> extends Collector<Row> {
+        /** */
+        private final List<Object> outBuf;
+
+        /** */
+        private ArrayCollector(
+            RowHandler<Row> handler,
+            RowHandler.RowFactory<Row> rowFactory,
+            int cap
+        ) {
+            super(handler, rowFactory, cap);
+            outBuf = new ArrayList<>(cap);
+        }
+
+        /** {@inheritDoc} */
+        @Override protected Object outData() {
+            return Collections.unmodifiableList(outBuf);
+        }
+
+        /** {@inheritDoc} */
+        @Override public void push(Row row) {
+            outBuf.add(rowHandler.get(0, row));
+        }
+
+        /** {@inheritDoc} */
+        @Override public void clear() {
+            outBuf.clear();
```

Comment:
I think it's not correct to return unmodifiable collection over the same instance after clear. `outBuf` should be recreated on each `clear()` call.
I've tried to write reproducer with CNLJ, but unfortunately found some other errors during planning. Something like this should reproduce the problem:
```
query T
SELECT ARRAY(SELECT b FROM t1 t1_1 WHERE t1_1.a = t1_2.a) FROM t1 t1_2 GROUP BY a ORDER BY a
----
[10, 22]
[21]
```
This reproducer is also not working since passThroughTraits/deriveTraits is not implemented for IgniteCollect:
```
query T
SELECT ARRAY(SELECT b FROM t1 t1_1 WHERE t1_1.a = t1_2.a) FROM t1 t1_2 ORDER BY a
----
[10, 22]
[10, 22]
[21]
```